### PR TITLE
fix(daemon): strip ANTHROPIC_API_KEY when spawning Claude Code

### LINKED ISSUE #398: Claude Code charges API usage despite "local CLI" auth setup with Pro account
I'm using Claude Code with a Claude Pro subscription. During setup, I selected "Local CLI" as the authentication method (see screenshots), expecting usage to be covered by my Pro plan rather than billed against an API key.

However, I'm still seeing charges for API usage on my account.

Steps to reproduce

    Set up Claude Code with a Pro account
    Choose "Local CLI" as the auth method during setup
    Use Claude Code normally
    Observe API usage charges accruing

What I've tried

    Unsetting the ANTHROPIC_API_KEY environment variable
    Confirming via /login (or settings) that "Local CLI" is the selected auth method

API usage continues to be billed regardless.

Expected behavior

With "Local CLI" selected and a Pro subscription, usage should be drawn from the Pro plan quota and not generate API charges.

Actual behavior

Usage is being billed as API usage (see third screenshot).

Environment

    Auth method: Local CLI
    Plan: Claude Pro
    ANTHROPIC_API_KEY: unset

Question

Is this expected behavior, or is there an additional step required to ensure Claude Code uses the Pro subscription instead of API billing?

Screenshots

Image
Image

Image

### PR DESCRIPTION
**Pull request by sebastianwestberg**
## Summary

- Fixes #398 — when the daemon spawns Claude Code, it inherits `ANTHROPIC_API_KEY` from the launching shell. Claude Code's [documented behavior](https://support.claude.com/en/articles/11145838-using-claude-code-with-your-pro-or-max-plan) is to prefer that env var over the user's `claude login` subscription credentials, so users on Pro/Max who selected "Local CLI" silently got billed against their API key instead.
- Strip `ANTHROPIC_API_KEY` from the env passed to `spawn` for the `claude` adapter only, so Claude Code's own auth resolution applies. Other adapters (which may legitimately use the var, e.g. when routing through Anthropic) are unaffected.

## Test plan

- [x] `pnpm --filter @open-design/daemon typecheck`
- [x] `pnpm --filter @open-design/daemon test` (342 tests pass)
- [x] `pnpm test` (full workspace) passes
- [x] `pnpm check:residual-js` passes
- [ ] Manual: with `ANTHROPIC_API_KEY` exported in the shell that launches the daemon, run a Claude Code chat and confirm the request hits the subscription rather than the API key (visible in Anthropic billing dashboard / `claude` debug logs)

### CODE REVIEW

**Comment by lefarcen** [comment]
Hi @sebastianwestberg! 🎉
Thanks for the contribution — excellent catch on the API key inheritance issue.
I will run a deep review and get back to you within 24h.

Thanks for making open-design better!
— open-design team

File: apps/daemon/src/server.ts [hunk 1]
```diff
@@ -2439,6 +2439,17 @@ export async function startServer({ port = 7456, host = process.env.OD_BIND_HOST
...
+      // Claude Code prefers ANTHROPIC_API_KEY over the user's logged-in
+      // subscription auth when both are present, silently billing API
+      // usage even though the daemon's "Local CLI" mode implies the user
+      // wants their `claude login` credentials (Pro/Max plan) to win. If
+      // the launching shell exported ANTHROPIC_API_KEY for any other
+      // reason (SDK work, scripts), the spawn would inherit it. Strip it
+      // for the claude adapter so the CLI's own auth resolution applies.
+      // See https://github.com/nexu-io/open-design/issues/398
+      if (def.id === 'claude') {
+        delete env.ANTHROPIC_API_KEY;
```
Comment (mrcfps) [inline apps/daemon/src/server.ts]: This strips only the exact uppercase key, so a mixed/lower-case `Anthropic_Api_Key` / `anthropic_api_key` entry can still be forwarded to `spawn` on Windows, where environment variable names are case-insensitive. That leaves the Claude adapter using the API key despite the Local CLI path intending to fall back to `claude login`, and it is especially easy to miss because spreading `process.env` into a plain object loses any case-insensitive lookup behavior before this `delete`. Could we remove this case-insensitively, e.g. iterate `Object.keys(env)` and delete every key whose `toUpperCase() === 'ANTHROPIC_API_KEY'` when `def.id === 'claude'`? A small helper/test around the spawned env shape with a mixed-case key would lock in the intended protection across platforms. 😊

**Review by mrcfps — COMMENTED** [review]
Thanks for the targeted security/UX fix here. I found one cross-platform edge case where the secret can still leak through on Windows because environment variable names are case-insensitive there.


<sub>Generated by Looper 0.0.0-dev · runner=reviewer · agent=opencode</sub>

File: apps/daemon/src/server.ts (hunk 1 above)
Comment (sebastianwestberg) [reply apps/daemon/src/server.ts]: Good catch, I confirmed locally that `{...process.env}` produces a plain object that loses Node's Windows case-insensitive accessor (a `Path` key stays `Path`), so a mixed-case `Anthropic_Api_Key` would have survived the literal `delete` and still been visible to `claude.exe` via `GetEnvironmentVariable`.

Pushed [9d8e68d](https://github.com/nexu-io/open-design/pull/400/commits/9d8e68d) which:

- Extracts the strip into a `spawnEnvForAgent(agentId, baseEnv)` helper in `agents.ts` next to `getAgentDef` / `resolveAgentBin`.
- Iterates `Object.keys(env)` and deletes any key whose `.toUpperCase() === 'ANTHROPIC_API_KEY'`.
- Adds four tests in `apps/daemon/tests/agents.test.ts`: uppercase strip for `claude`, case-insensitive strip (mixed and lower) for `claude`, preservation for non-claude adapters (`codex` / `gemini` / `opencode` / `devin`), and input immutability.

`pnpm --filter @open-design/daemon typecheck` and `pnpm --filter @open-design/daemon test` both clean (346 tests pass, +4 from before).

**Review by mrcfps — APPROVED** [review]
@sebastianwestberg I reviewed the scoped daemon changes: `spawnEnvForAgent` now copies the base env, strips every case variant of `ANTHROPIC_API_KEY` only for the Claude adapter, preserves the key for other adapters, and the added unit coverage locks in the strip, preservation, and immutability behavior. No actionable issues found in the changed ranges — thanks for turning the previous cross-platform edge case into a focused helper with clear tests. 😊


<sub>Generated by Looper 0.0.0-dev · runner=reviewer · agent=opencode</sub>

**Review by lefarcen — COMMENTED** [review]
Thanks for the quick, thorough fix @sebastianwestberg! 🎉

The case-insensitive deletion + comprehensive test coverage fully addresses the Windows leak mrcfps identified. The helper extraction is clean and the tests lock in the security boundary across platforms.

Looks good to me; deferring final approval to a maintainer.

**Comment by lefarcen** [comment]
Merged! 🎉 Thanks @sebastianwestberg for the contribution to open-design. Looking forward to the next one!

**Comment by lefarcen** [comment]
Merged! 🎉 Thanks @sebastianwestberg for the contribution to open-design. Looking forward to the next one!
